### ISSUE TITLE
Help Needed with Migrating Dialog

### ISSUE THREAD

**Issue by Harbs** [adobe/spectrum-css#892]
## Description
The layout of Dialog completely changed and I'm not sure how to use the new layout.

I have an image above the dialog header and I don't see how that can work using the new css. The grid layout forces everything into very rigid predefined positions. I need the flexibility of being able to customize things as necessary.

Suggestions?

After playing around, it seems like I can wrap elements in something like this: 
```
<section class="spectrum-Dialog-heading">
<div><img/><p></p><div/>
</section>
````
Is that the suggested way to go about it?

I'm also confused on what determines whether buttons go at the top or the bottom of the dialog. How is that supposed to be controlled?

**Comment by misterbrownlee** [comment]
FYI, we got this request and are chatting about it before responding.  Thanks!

**Comment by Harbs** [comment]
Working on this more, I ran into a big problem.

It looks like the buttons are positioned on the bottom unless it's a full screen dialog. The the grid places the buttons at the top right.

Well, that's not going to work very well for me. I use full screen dialogs extensively in CEP panels because that's the best I have to work with there. They are "normal" dialogs and the action buttons belong at the bottom of the dialog and not the top.

I need a solution to this problem.

**Comment by Harbs** [comment]
(Slight clarification: I use full screen takeover dialogs in my panels, but the point is the same.)

**Comment by GarthDB** [comment]
@Harbs the design language doesn't support the [full screen dialogs](https://spectrum.adobe.com/page/dialog/). It's possible we will remove them, or they could be redesigned completely before the are fully supported.

**Comment by Harbs** [comment]
They were working perfectly before the layout was changed from flexbox for grid.

They are very important for CEP panels.

**Comment by Harbs** [comment]
I'm trying to understand the reasoning of switching dialog from flexbox to grid. Grid is very cool when it solves a problem, but I'm not sure what problem needed solving with dialog.

The downsides of grid are:
1. It makes dialogs much more difficult to customize beyond the standard uses.
2. The CSS is more verbose.
3. AFAIK `grid-template-areas` does not work in IE. I'm not sure I understand how this all displays in IE11.

Basically, it transfers the "structure" of the dialog from HTML to CSS. I'm not sure why that's desirable. I'd much rather have the structure controlled by the HTML I'm generating.

**Comment by GarthDB** [comment]
@Harbs

1. This has been implemented after discussions with @snowystinger on [React Spectrum](https://github.com/adobe/react-spectrum). It actually makes rearranging elements easier and more semantic.
1. It does make the CSS more verbose, but I think the semantics and flexibility worth it.
1. We are dropping support for IE11.

**Comment by Harbs** [comment]
> It actually makes rearranging elements easier and more semantic.
Can you explain how? Arranging via CSS, maybe. Arranging via HTML? I can't see how.

How as a consumer am I supposed to rearrange elements?

The image in the header (pushing down the text) is one example. The buttons on the top/bottom is another. I'm not sure why the "sticky footer" was dropped.

> We are dropping support for IE11.

I need to figure out if this is a problem for me...

> This has been implemented after discussions with @snowystinger on React Spectrum.

Can I suggest that more of these discussions happen in the open? I'd love to participate and I likely have a perspective different than what you have internally.

**Comment by Harbs** [comment]
> We are dropping support for IE11.
What does that mean for the color stop builds? AIUI, there's two reasons to create the color stop builds. One is IE support and the second is support for CEP7. Is there another reason why color stops are important (rather than using css vars)?

**Comment by snowystinger** [comment]
The above is a good summary of why the changes were made

Dialogs have a very well defined layout right now, see https://spectrum.adobe.com/page/dialog/ and https://spectrum-contributions.corp.adobe.com/page/takeover-dialog-beta/ 

We recognize that these patterns will evolve and may change, as such CSS Grid gives us the flexibility to reorder the placement of groups of components without changing the underlaying DOM structure. This means we can make non-breaking change whenever the design updates (to a large extent). Note* this is not to make rearranging easier for the consumer, this is to make it easier for Design to do updates and for our (and spectrum-css) library to be non-breaking.
Also, because this is just CSS, there is the option to override pretty much all of it. It is very much discouraged because there is a Spectrum pattern that should be followed as much as possible.
See https://react-spectrum.adobe.com/react-spectrum/Dialog.html#props

Finally, you are capable of creating your own Dialogs if the support for Grid is really that limited. See https://github.com/adobe/react-spectrum/blob/main/packages/%40react-spectrum/dialog/src/Dialog.tsx, this file is just about the rendering of Dialogs, you could swap out that portion and still use all the hooks for behavior.
We do use grid in other places though because it is so powerful in allowing us to make large design changes with no updates to the actual DOM structure, just something to keep in mind. If you aren't using React Spectrum, then you can just use the classnames you need and change the display property as you need.

To a large degree, we did attempt to have this discussion in the open. Here's the RFC for Slots in v3 https://github.com/adobe/react-spectrum/blob/main/rfcs/2019-v3-slots.md which is tightly tied to the usage of CSS Grid.
And here's the RFC for theming, which is made possible with CSS variables and also necessitated dropping IE11 support. https://github.com/adobe/react-spectrum/blob/dbf4b55167c974077ae6f6b0347a2a1b3947a91f/rfcs/2019-v3-theming.md#native-css-variables
Though I can see how this would be less transparent if you are not a user of React Spectrum.

We would welcome your perspective and have a number of alpha components for React Spectrum v3. You can raise issues for anything you find at https://github.com/adobe/react-spectrum/issues. We also accept RFC's to discuss the architecture. 
@GarthDB  will need to chime in again if it's specific to Spectrum CSS and how they would like to have these discussions.

**Comment by Harbs** [comment]
Hi @snowystinger Thanks for taking the time to respond.

We are in fact not using React. I created a pretty complete component set which uses the spectrum css:
https://github.com/unhurdle/spectrum-royale
https://unhurdle.github.io/spectrum-royale/

I'm currently working on updating it and expanding the support.

My preferred development environment is [Royale](https://royale.apache.org/) and I've seen interest in using spectrum css with Vue, Svelte and other.

I have no access to the corp.adobe.com page, so I'm not sure what the current status of the takeover dialogs are.

To be clear, I had no issues with rendering Dialog using the current css. https://github.com/unhurdle/spectrum-royale/commit/b8d665bbf87b4d0021bcf493bfa24559f465e577 My issue is simply that the current css does not support my use cases for the existing dialog component that I have in production products. That includes a supported way to customize the header and footer areas of the dialog. Footers seem to be supported in normal dialogs, but not fullscreen/takeover ones.

I considered trying a hacky solution for positioning buttons at the footer of a takeover dialog, but I couldn't come up with something. I'm pretty new to grid, so it's possible I'm missing something there, but I have no idea how to override grid template areas.

I hear your points on grid being useful for redesigns, but to be honest I'm kinda nervous about the concept of my apps using Spectrum to suddenly take on a new layout with a css change. Despite the best efforts to include a full design description in a css implementation, I've had to make many local css overrides in practice. This is especially true for CEP panels where we need to fit design elements into smaller-than-planned spaces...

**Comment by Harbs** [comment]
> We are dropping support for IE11.

Yes. This is a problem for me.

We had a Japanese client tell us as recently as April of this year that IE11 is a requirement.

In the last day alone, we've had at least 10 clients use our app in IE.

I know supporting IE is a pain, but I don't think I'm at the point where I can drop support for it.

**Comment by GarthDB** [comment]
@Harbs have you considered locking to a version of Spectrum CSS that has what you need? Is there something specific you're hoping to get from the future updates?

**Comment by GarthDB** [comment]
@harbs, I don't think there is anything on the internal adobe sites that aren't on the public ones. There is no discussion on Dialog currently with the design team.

**Comment by Harbs** [comment]
> have you considered locking to a version of Spectrum CSS that has what you need? 

I'm considering doing just that.

> Is there something specific you're hoping to get from the future updates?

Other than the components that need finishing? Probably not.

Until recently the color components were kind of raw.
The calendar/date picker interaction was not completely defined. Not sure where that stands now.
Off the top of my head, I'm not sure what else I'd be looking for.

My comment about the corp site was about @snowystinger link to a takeover-dialog-beta page.

Thanks.

**Comment by snowystinger** [comment]
Ah, sorry, I had assumed you were an Adobe employee, no worries. When fullscreen gets out of contribution phase, it'll make its way into the public documentation.

As for using spectrum css with other frameworks, we're really hoping that we can push people towards something that is kindred to React Spectrum. See https://github.com/adobe/react-spectrum/issues/995 where we've started trying our architecture with say Svelte. React Spectrum is architected in a way to try to share behaviors and accessibility and leave the rendering layer out of it. Feel free to chime in there.
As for why we're doing that, you might find this to be an interesting read https://react-spectrum.adobe.com/blog/building-a-button-part-1.html.

I get that you don't want your visual design to change out from under you, however, as long as all parts are maintained and accessible, it shouldn't matter too much and would help stay in line with Adobe's products which are also largely using React Spectrum. The use of CSS Grid allows us to easily maintain the same tabbing order even with a new visual layout. I don't expect the design to change drastically with much frequency, and products can always pin to a specific version until they are ready for the new look.

If you don't want Spectrum's look, then creating your own rendering shouldn't be too hard, there's examples using our hooks in https://react-spectrum.adobe.com/react-aria/index.html or if you don't need something so extreme, you can check out theming https://react-spectrum.adobe.com/react-spectrum/theming.html

**Comment by GarthDB** [comment]
@Harbs we are also releasing components individually now, so you might be able to grab some of those components à la carte.

**Comment by GarthDB** [comment]
@Harbs a big reason we are making a design system is so in the future, if there are updates, they are easy to propagate throughout all the products using Spectrum.

**Comment by Harbs** [comment]
> As for using spectrum css with other frameworks, we're really hoping that we can push people towards something that is kindred to React Spectrum.

I hear you, and that might be possible with Vue or Svelte, but I don't think it's going to work very well with Royale.

I'm in awe of the work that @devongovett has done both here and elsewhere (i.e. fontkit). He probably has an IQ an order of a magnitude higher than mine. 😉 I'm generally adverse to magic and architecture which makes such magic happen. If I don't understand how something works, I have difficulty using it. I did download the React repo and browsed it. Reading through the react packages made my eyes water... 😉 (Although a second browse-through was easier.) The blog post was very interesting. It's a great read.

FWIW, I believe the need for such complicated architecture is vastly reduced in Royale. I find Royale architecture very straight-forward which is a large part of my attraction to it. I definitely plan on implementing much of what was done in react-aria. I'm not sure how much of react-stately is even needed in Royale where state is preserved in components. Royale does have the concept of injecting behavior and that's done by something we call "beads".

If you're interested, I just recently added basic focus management that is not component specific here: https://github.com/unhurdle/spectrum-royale/commit/304f403aa22df50f842975bbd854c8a04aeb4ab0 To subscribe, a component implements `IKeyboardFocusable` and just adds the `KeyboardFocusHandler` bead (either declaratively or imperatively). I do believe that it's lighter-weight than the React implementation although I'm not sure if I covered all the bases yet. It seems to work well.

**Comment by snowystinger** [comment]
> I'm in awe of the work that @devongovett has done

you and me both :)

I did leave out one more reason for CSS Grid, Spectrum-CSS is improving its support for RTL and mobile responsive designs. This is largely due to the use of CSS Grid. RTL comes almost free as Grid will just change the direction of the inline axis. Responsive is made easier since the same DOM structure can be visually expressed in many different ways.
In the case of fullscreen dialogs in a large screen format, we can spread out the header elements horizontally and there is room for the button group removing the need for a footer. In a mobile setting though, we want those buttons to be easy to reach with a thumb, so we need them to be moved below the content. This cannot be expressed in CSS Flex or any other way without changes to the DOM which would put a burden on developers using our components and would make them less flexible. (I realize the pure example could be expressed that way, but there are enough other items in a dialog that it really can't be)

**Comment by Harbs** [comment]
RTL is a very good point. Considering I'm very RTL oriented I don't need to be convinced there.
